### ISSUE TITLE
Rework `_GenerateJavaDesignerForComponent`

### ISSUE THREAD

**Issue by dellis1972** [xamarin/xamarin-android#2836]
The `_GenerateJavaDesignerForComponent` is tasked with generating the `R.java` files for all the referenced libraries which have an `AndroidManifest.xml`. This is so any java code which use resource items such as `android.support.transition.R.abc_fade_in` will compile. 

Currently we call `aapt` or `aapt2` once per library to generate the file. We do this on the first build. But we also do it on subsequent builds if the library changes. For Nuget packages this means once its built it never gets built again unless the user does a `Clean`.

We do run these tasks in Parallel, which does improve things quite a bit. But it does put the CPU under allot of stress, since we have to launch a `aapt*` process *per* library. So if you have 8 cores on your machine you get 7 `aapt*` processes running.

Is this the most efficient way of doing this? Looking at the output of the `R.java` file the ONLY code difference between ALL the files is the package declaration at the top.
E.g `package android.support.compat;` vs `package android.support.transition;`
So the next question is why don't we generate one `R.java` file and then do a search & replace for each library reading the required `packageName` from the `AndroidManifest.xml`.

This would mean 

A) We are not duplicating a whole TON of work
B) the initial build should be much quicker since we are not spawning up a whole ton of processes.

**Comment by jonathanpeppers** [comment]
So if I look at the `<Aapt2Link/>` call for one of these:

```
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\10\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp944D.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\13\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp945F.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\22\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp9473.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\15\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp944E.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\18\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp9462.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\20\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp945E.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\19\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp9461.tmp
    Executing link --manifest C:\Users\jopepper\Desktop\Git\xamarin-android\tests\Xamarin.Forms-Performance-Integration\Droid\obj\Debug\90\lp\16\jl\res\..\manifest\AndroidManifest.xml --java obj\Debug\90\android\src -R obj\Debug\90\flata\a4b85b611459d0952d6a8dfe9ad015c40a882909.flata -R obj\Debug\90\flata\d8ab1a034f0be43229b4da43251e84f2f7e9b249.flata -R obj\Debug\90\flata\3fff511ef9264bad797ba8942541ecffd454b351.flata -R obj\Debug\90\flata\5f8a75baa4723a993fc610218c21ecf3b47a1c37.flata -R obj\Debug\90\flata\e8c79609a2a8bfa4c540009f80389af291132eea.flata -R obj\Debug\90\flata\796f9008b9fc36582ee0579e7506dea4d7cc0686.flata -R obj\Debug\90\flata\f6ec8bb92c8420670fff18f26c063c37d80eca24.flata -R obj\Debug\90\flata\f2bead6468118783c404222da88d7e9e9ede701f.flata -R obj\Debug\90\flata\\compiled.flata --auto-add-overlay -I C:\Users\jopepper\android-toolchain\sdk\platforms\android-28\android.jar -o C:\Users\jopepper\AppData\Local\Temp\tmp9460.tmp
```

It looks like this switch is putting `R.java` files per Java package to be built and placed in the app:
```
--java obj\Debug\90\android\src
```

If we built a single `R.java`, how would we know how to split up the massive file into multiple files? I would think the resource ids would be all mixed up and you wouldn't know which one came from which package?

I think there can be Java code that depends on `R.id.foo` working via a specific package name.

**Comment by dellis1972** [comment]
We use the list of `AndroidManifest.xml` files to drive where the `R.java` files need to be copied too. 
Note that ALL the `R.java` files are the same they contain ALL the resources the only difference is the `package` value and a the comments. The values of the id's are irrelevant since the final app `R.java` replaces them all.

**Comment by jonathanpeppers** [comment]
Ok, I see:
* So we _should_ be creating an `R.java` that only contains resources for a given jar file (and any other jars _it_ references)
* We are actually creating `R.java` files that contain _all resources_ for all jar files...

Since we are already doing things this way, we can reduce this to a single `aapt` call.

The only other diffs I see are in the `javadoc` comments (don't matter), but I looked at `aapt2` output. I will check `aapt` to be sure it's the same.

**Comment by jonathanpeppers** [comment]
Ok, I think our current behavior actually might be a problem:
> It has been observed that Xamarin.Android apps "seem" to hit the dex limit sooner than Java apps.

I have not had time to actually investigate and see if this is the case--and see by _how much_.

For a "Hello World" Xamarin.Forms app, if I look at `R.java`, it contains 2,267 fields. There are 9 `R.java`'s: 8 for libraries, and 1 for the main app.

Does this mean we have potentially some amount of extra fields? Could be some percentage of 18,136 fields (8 x 2,267)?

**Comment by jonathanpeppers** [comment]
Ok so if I just look at a random class: `Landroid/support/compat/R$drawable;`

* [Google Play Service Java Sample](https://github.com/googlemaps/android-samples/tree/master/ApiDemos/java): 11 fields
* Xamarin.Forms "Hello World": 310 fields

Here is dexdump output of each: [dexdump.zip](https://github.com/xamarin/xamarin-android/files/3005414/dexdump.zip)

**Comment by dellis1972** [comment]
The problem we have is there are java classes in the support libs which reference `R.xx` values for other libraries. For example `AppCompat` Theme stuff is referenced and used by other libraries and `aar` files. During compilation if javac cannot find these we get a build error. These are the build errors we see when Nuget's have not been restored properly. 

So I believe that the whole point of `_GenerateJavaDesignerForComponent` was to generate those `R.java` files for libraries. Now I would love to cut down on what we generate but we don't really know what is being referenced until we compile the code. It does feel like this target (`_GenerateJavaDesignerForComponent`) is a bit of a hammer for fixing a particular problem, and probably not in the most efficient way.

**Comment by jonathanpeppers** [comment]
Ok, so I think I understand now what Java/Android Studio does.

At the top level, the Android gradle plugin has a [`processResources` task](https://android.googlesource.com/platform/tools/build/+/refs/heads/master/builder/src/main/java/com/android/builder/AndroidBuilder.java#599).

It runs `aapt` once on a single directory to produce an `R.txt`. All their resources must be dumped into a single directory.

Then the `R.txt` file is [parsed via SymbolLoader.java](https://android.googlesource.com/platform/tools/build/+/refs/heads/master/builder/src/main/java/com/android/builder/internal/SymbolLoader.java) and only fields that are needed are included for each library: https://android.googlesource.com/platform/tools/build/+/refs/heads/master/builder/src/main/java/com/android/builder/AndroidBuilder.java#726

This Java code writes out the individual `R.java` files via [SymbolWriter.java](https://android.googlesource.com/platform/tools/build/+/refs/heads/master/builder/src/main/java/com/android/builder/internal/SymbolWriter.java).

So we would have to port this code to C#? Or consider integrating with `gradle`?

**Comment by dellis1972** [comment]
lets not integrate with gradle if we can help it. We already produce the R.txt with our calls to `aapt` and `aapt2`. So we could just add this as a step before CompileJava and replace the `_GenerateJavaDesignerForComponent`

**Comment by dellis1972** [comment]
We have a basic `R.txt` parser in the ManagedResourceParser https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Utilities/ManagedResourceParser.cs#L159.

my App `R.txt` for a Standard Xamarin.Forms app has 2900 entires in it. 

Looking at the code you posted it seems the `AndroidBuilder.java` uses both the main app `R.txt` AND the `R.txt` which is included in the library. 
https://android.googlesource.com/platform/tools/build/+/refs/heads/master/builder/src/main/java/com/android/builder/AndroidBuilder.java#745

Luckily we do include that in our extraction process so we have `obj/Debug/lp/32/R.txt` for example. These files are much smaller (or don't exist). 

So I guess they do some kind of merge process. between that main one and the one the library has.

**Comment by erikpowa** [comment]
> This is so any java code which use resource items such as android.support.transition.R.abc_fade_in will compile.

For the record, the other side acting the same way with `Resource.designer.cs`. If you have a XA Library referencing foreg the appcompat library and uses **1** resource from it, there will be `Resource.designer.cs` with junk static fields and field setters for the `UpdateValues`. Problem/bad part is the Linker can't just catch them all (the static array initialization are the annoying ones).

Since I have OCD against junk IL code, I try to move resource access out from XA Libraries, to the root App via XML + IDs (this way the `Resource.designer.cs` will not have the the appcompat resource ids, only the ones I specify) in XA libraries and to ensure things, I do pre-link before LinkAssemblies, looking for unused 
 resource access aka for (resource)fields that doesn't have `ldsfld` anywhere, then clean up any `stsfld` (references/patterns like `ldc_i4 stsfld` or `ldc_i4_1 newarr dup ldtoken call stsfld`) and then just removing the fields themselves.)

**Comment by jonathanpeppers** [comment]
@erikpowa if you enable the linker (link all assemblies) are these removed? You can also add `[assembly:LinkerSafe]` if you want the SDK only option to strip them.

You can also prevent `Resources.designer.cs` from being created at all in a library project if you just omit `<AndroidResgenFile>Resources\Resource.Designer.cs</AndroidResgenFile>`. Xamarin.Forms does this (or excludes the file from `<Compile/>`) for several of their library projects that reference the support libraries.

**Comment by erikpowa** [comment]
@jonathanpeppers 

> if you enable the linker (link all assemblies) are these removed?

Not really... especially from the `RootAssembly` (as far as I know the root assembly being **skip**ped by default and that holds basically every resource id). About the libraries, linker strips only the classes like `Dimension`  **if the class itself is unused**. I'm pretty sure stripping the resources will require a custom `Step` (if possible) for the Linker, because static constructor initializes static fields and if that class being called / used in _anyway_, that `.cctor()` will stay 100% with it's field setter opcodes, therefore the fields are staying aswell because the linker think they are used (tested it long ago). That's why I wrote a custom link task with few others.


> [assembly:LinkerSafe]

I never heard of it actually. Forgot to mention that my context is with `AndroidLinkMode=Full`.

> if you just omit 

Well if I omit it, how would that compile if I have resource that needs to be accessed? or how the resource id would point to the correct resource ? since the `RootAssembly` should contain that static `UpdateIdValues()` within that the updates for those static fields to the correct resource ids.

Btw the `UpdateIdValues()` in the `RootAssembly` will contain `ldsfld stsfld` when static field updates static field... remember I said that if the resource class is being used, then the whole class stay with all it's field... + the `RootAssembly` not stripped = lot's of unused IL code everywhere

**Comment by jonathanpeppers** [comment]
@erikpowa so I think we should revisit the managed side after fixing the java side.

Either a new linker step that strips unused fields--or maybe we don't need to generate as much C# code from being generated in the first place? Although, for Intellisense to work, it would need fields that aren't in use yet...

If you have a linker step fixing this up, is it compatible with our stuff? would you be willing to open a PR?

**Comment by erikpowa** [comment]
@jonathanpeppers I haven't dived myself into the designtime resource-generator, but the only problem is the **"strong references" to the fields**. As I see as for solution:
Currently, library resource fields are required to be static, because `UpdateIdValues()` will change them when app starts (or atleast when the Resource class first accessed, which is also bad because of it uses reflection when "_we know the metadata at compile time_"). Foreg:
`global::CommonUI.Resource.Dimension.size200 = global::MainAssembly.Resource.Dimension.size200;`
this is one line from `UpdateIdValues()` in `MainAssembly`
where `global::CommonUI.Resource.Dimension.size200` is static field from a library
where `global::MainAssembly.Resource.Dimension.size200` const field from the `RootAssembly` (having the final resourceId)

If the res designer already knows what resource will be set for which static field, then why isn't this knowledge/process moved out?
Let's say `global::MainAssembly.Resource.Dimension.size200` has value `2131099880`. 
"LinkAssemblies" uses mono cecil anyway, it would be easier just modify the field's value in the libraries when the app compiles (we know the field, we know what value it's gonna get) Foreg: In `CommonUI` library, the `global::CommonUI.Resource.Dimension.size200` will be set to `2131099880` in the `UpdateIdValues()`, but if we could make the `global::CommonUI.Resource.Dimension.size200` a const field and modify the IL code to have a value with `2131099880` then all the problem will disappear... no longer strong reference to fields, no longer required to call `UpdateIdValues()`, no static constructor with trigger of `UpdateIdValues`, no reflection used and the linker will remove any unused const field(?).

> is it compatible with our stuff

For resource cleaning, I have just temporary solution, unsafe and not optimized at all.
[Here](https://gist.github.com/erikpowa/106c3d9b160e43a1497cf796aeeffe28)

**Comment by jonathanpeppers** [comment]
@erikpowa to add a new linker step, you would probably have to add a new linker step to `Xamarin.Android.Build.Tasks.dll` and build your own version. See: https://github.com/xamarin/xamarin-android/blob/e6326dcba43dfb02c7401de8ae38c03166505df8/src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/Linker.cs#L58

I'll file a new issue about this--it is kind of unrelated to the `R.java` issue we are trying to solve here.
